runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/SdkHarnessRegistries.java

```diff
@@ -107,6 +108,17 @@ public void unregisterWorkerClient(FnApiControlClient controlClient) {
         workers.remove(worker);
       }
       LOG.info("Unregistered Control client {}", worker != null ? worker.getWorkerId() : null);
+
+      // unregisterWorkerClient() will be called only when the connection between SDK harness and
+      // runner harness is broken or SDK harness respond to runner harness with an error. In either
+      // case, the SDK should be marked as unhealthy.
+      sdkHarnessesAreHealthy.set(false);
```

Comment:
Is this the only place that will catch sdk harness issues? Would this work for all cases?

Reply:
This doesn't work for all cases now. It only covers cases where the grpc channel between SDK harness and runner harness is broken/erroneous.

I am looking at what changes needed to be added to the SDK harness side now (This PR only covers changes on the runner harness side), in order to make it cover more cases. I plan to send another PR to do that, or patch this PR if there aren't many changes needed.

Reply:
That sounds good. Maybe add a comment on the bug or here with a comment on what are the limitations.

Previously we had a conversation about not receiving progress updates as an indicator of sdk harness failure. We could consider that as an option.

Reply:
Taking progress update into consideration as an indicator of SDK harness health is definitely something worth exploring, but before that I want to focus the current discussion (see the bug) on whether we should treat bundle execution failure as SDK harness unhealthy. If not, what do we need to do if the bundle keeps failing? Do we simply retry indefinitely locally?

Reply:
Ack. Let's keep the discussion in one place. I will comment in the other place.